core/blockchain/blockchain_gateway.py

```diff
@@ -0,0 +1,355 @@
+        fulfilled on a per tx basis
+        `poll_interval` specifies the number of seconds to stall before each
+        poll
+        """
+        while True:
+            filtered_diffs = self.get_state_diffs(event_filter)
+            if filtered_diffs:
+                return filtered_diffs
+            await asyncio.sleep(poll_interval)
+
+    def filter_set(self, event_filter, handler):
+        """
+        Calls async method `start_listening` and called by various listening
+        methods
+        """
+        asyncio.set_event_loop(asyncio.new_event_loop())
+        loop = asyncio.get_event_loop()
+        try:
+            filtered_diffs = loop.run_until_complete(
+                self.start_listening(event_filter, handler))
+            check = [handler(diff) for diff in filtered_diffs]
+        finally:
+            loop.close()
+        return check
+
+    def handle_decentralized_learning_trainer(self, tx: dict) -> None:
+        """
+        This callback will be triggered by the Listener if it finds the 
+        method signature it's looking for.
+        The parameters (model weights, model config) will be downloaded 
+        and put into the optimizer initially. So the optimizer knows this info.
+        """
+        logging.info("handling decentralized learning... {}".format(tx))
+        key = tx.get('key')
+        value = tx.get('content')
+        args = {'key': key, 'content': self._ipfs_to_content(value)}
+        self.communication_manager.inform("new_session", args)
+        self.listen_new_weights()
+
+    def handle_new_weights(self, tx: dict):
+        """
+        handle_new_weights() method downloads weights and does something with
+        it. This callback is triggered by the Listener when it sees new weights 
+        intended for its node ID. Downloads the weights, and gives them to the
+        comm. mgr which gives them to the relevant optimizer 
+        -which should do the moving average.
+        """
+        logging.info("handling new weights...{}".format(tx))
+        key = tx.get('key')
+        value = tx.get('content')
+        args = {'key': key, 'content': self._ipfs_to_content(value)}
+        # TODO: Put into in-memory datastore.   
+        self.communication_manager.inform("new_weights", args)
+        self.listen_new_weights()
+
+    def handle_terminate(self):
+        self.communication_manager.inform("TERMINATE", None)
+
+    def listen_decentralized_learning(self):
+        """
+        Polls blockchain for node ID in decentralized_learning() method
+        signature decentralized_learning(...<node_ids>...) method signature
+        will be the key on the blockchain; listener should look for this, and
+        if the method signature contains its node id, it will trigger a
+        callback
+        """
+        def filter(tx):
+            logging.info("tx: {}".format(tx))
+            return tx.get('key') == tx.get('content')
```

Comment:
Same as above.

Reply:
same as above

Reply:
@nzoghb same as above

Reply:
asserts would work (but also please note that if these functions are only used by other functions that already do the assert, you don't need to do it on every function -- just add the precondition on the docstring)